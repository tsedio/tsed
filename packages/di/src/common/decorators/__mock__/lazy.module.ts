import {Injectable} from "../injectable.js";

@Injectable()
export class MyLazyModule {
  called = false;

  $onInit() {
    this.called = true;
  }
}

export default MyLazyModule;
