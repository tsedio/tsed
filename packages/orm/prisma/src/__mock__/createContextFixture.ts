import {TransformContext} from "../generator/domain/TransformContext.js";

export function createContextFixture(): TransformContext {
  return {
    modelsMap: new Map()
      .set("User", {
        name: "User",
        fields: [{type: "Role"}]
      })
      .set("Role", {
        name: "Role",
        fields: [{type: "User"}, {type: "Transitive"}]
      })
      .set("Transitive", {
        name: "Transitive",
        fields: [{type: "String"}, {type: "User"}]
      })
      .set("Hello", {
        name: "Hello",
        fields: [{type: "String"}, {type: "User"}]
      })
  } as any;
}
