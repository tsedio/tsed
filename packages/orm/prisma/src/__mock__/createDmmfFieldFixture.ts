import {DmmfField} from "../generator/domain/DmmfField";

export interface FieldFixtureOptions {
  kind: string;
  name: string;
  isRequired: boolean;
  isNullable: boolean;
  type: string;
  isList: boolean;
  documentation: string;
  isInputType: boolean;
}

export function createDmmfFieldFixture(options: Partial<FieldFixtureOptions> = {}) {
  return new DmmfField({
    field: {
      name: "user",
      isRequired: false,
      type: "User",
      isList: false,
      ...options
    },
    schemaArg: {
      ...options
    },
    model: {
      isInputType: options.isInputType,
      addImportDeclaration: jest.fn()
    }
  });
}
