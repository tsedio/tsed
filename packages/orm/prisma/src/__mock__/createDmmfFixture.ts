export function createDmmfFixture() {
  return require("./dmmf.json");
}

export function createDmmfWithTypesFixture() {
  return require("./dmmf.types.json");
}
