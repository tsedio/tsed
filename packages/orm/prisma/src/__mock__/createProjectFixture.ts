import filedirname from "filedirname";
import {join} from "path";
import {ModuleKind, Project, ScriptTarget} from "ts-morph";

// FIXME remove when esm is ready
const [, rootDir] = filedirname();

const SNAPSHOT_DIR = `${rootDir}/../../test/snapshots`;

export function createProjectFixture(dir = "/") {
  const baseDir = join(SNAPSHOT_DIR, dir);
  const project = new Project({
    compilerOptions: {
      target: ScriptTarget.ES2019,
      module: ModuleKind.CommonJS,
      emitDecoratorMetadata: true,
      experimentalDecorators: true,
      esModuleInterop: true
    }
  });

  return {
    project,
    baseDir,
    render(file: string) {
      const absolutePath = join(baseDir, file);
      const sourceFile = project.getSourceFile(absolutePath);

      if (!sourceFile) {
        throw `File not found: ${file}`;
      }

      sourceFile.formatText({indentSize: 2});
      const actualValue = sourceFile.getFullText();

      return {
        ...expect(actualValue),
        toMatchSnapshot() {
          expect(actualValue).toMatchSnapshot();
        }
      };
    }
  };
}
