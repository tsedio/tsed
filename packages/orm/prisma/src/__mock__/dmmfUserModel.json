{
  "model": {
    "name": "User",
    "isEmbedded": false,
    "dbName": null,
    "fields": [
      {
        "name": "id",
        "kind": "scalar",
        "isList": false,
        "isRequired": true,
        "isUnique": false,
        "isId": true,
        "isReadOnly": false,
        "type": "Int",
        "hasDefaultValue": true,
        "default": {
          "name": "autoincrement",
          "args": []
        },
        "isGenerated": false,
        "isUpdatedAt": false,
        "documentation": "@TsED.Groups(\"!creation\")\nComment"
      },
      {
        "name": "createdAt",
        "kind": "scalar",
        "isList": false,
        "isRequired": true,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "DateTime",
        "hasDefaultValue": true,
        "default": {
          "name": "now",
          "args": []
        },
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "email",
        "kind": "scalar",
        "isList": false,
        "isRequired": true,
        "isUnique": true,
        "isId": false,
        "isReadOnly": false,
        "type": "String",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false,
        "documentation": "@TsED.Email()\n@TsED.Description(\"User email. This email must be unique!\")"
      },
      {
        "name": "weight",
        "kind": "scalar",
        "isList": false,
        "isRequired": false,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "Float",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "is18",
        "kind": "scalar",
        "isList": false,
        "isRequired": false,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "Boolean",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "name",
        "kind": "scalar",
        "isList": false,
        "isRequired": false,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "String",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "successorId",
        "kind": "scalar",
        "isList": false,
        "isRequired": false,
        "isUnique": false,
        "isId": false,
        "isReadOnly": true,
        "type": "Int",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "successor",
        "kind": "object",
        "isList": false,
        "isRequired": false,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "User",
        "hasDefaultValue": false,
        "relationName": "BlogOwnerHistory",
        "relationFromFields": ["successorId"],
        "relationToFields": ["id"],
        "relationOnDelete": "NONE",
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "predecessor",
        "kind": "object",
        "isList": false,
        "isRequired": false,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "User",
        "hasDefaultValue": false,
        "relationName": "BlogOwnerHistory",
        "relationFromFields": [],
        "relationToFields": [],
        "relationOnDelete": "NONE",
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "role",
        "kind": "enum",
        "isList": false,
        "isRequired": true,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "Role",
        "hasDefaultValue": true,
        "default": "USER",
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "posts",
        "kind": "object",
        "isList": true,
        "isRequired": true,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "Post",
        "hasDefaultValue": false,
        "relationName": "PostToUser",
        "relationFromFields": [],
        "relationToFields": [],
        "relationOnDelete": "NONE",
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "keywords",
        "kind": "scalar",
        "isList": true,
        "isRequired": true,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "String",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "biography",
        "kind": "scalar",
        "isList": false,
        "isRequired": true,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "Json",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false
      }
    ],
    "isGenerated": false,
    "idFields": [],
    "uniqueFields": [],
    "uniqueIndexes": []
  },
  "modelType": {
    "name": "User",
    "fields": [
      {
        "name": "id",
        "args": [],
        "isNullable": false,
        "outputType": {
          "type": "Int",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "createdAt",
        "args": [],
        "isNullable": false,
        "outputType": {
          "type": "DateTime",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "email",
        "args": [],
        "isNullable": false,
        "outputType": {
          "type": "String",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "weight",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "Float",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "is18",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "Boolean",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "name",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "String",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "successorId",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "Int",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "successor",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "User",
          "namespace": "model",
          "location": "outputObjectTypes",
          "isList": false
        }
      },
      {
        "name": "predecessor",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "User",
          "namespace": "model",
          "location": "outputObjectTypes",
          "isList": false
        }
      },
      {
        "name": "role",
        "args": [],
        "isNullable": false,
        "outputType": {
          "type": "Role",
          "namespace": "model",
          "location": "enumTypes",
          "isList": false
        }
      },
      {
        "name": "posts",
        "args": [
          {
            "name": "where",
            "isRequired": false,
            "isNullable": false,
            "inputTypes": [
              {
                "type": "PostWhereInput",
                "namespace": "prisma",
                "location": "inputObjectTypes",
                "isList": false
              }
            ]
          },
          {
            "name": "orderBy",
            "isRequired": false,
            "isNullable": false,
            "inputTypes": [
              {
                "type": "PostOrderByWithRelationInput",
                "namespace": "prisma",
                "location": "inputObjectTypes",
                "isList": true
              },
              {
                "type": "PostOrderByWithRelationInput",
                "namespace": "prisma",
                "location": "inputObjectTypes",
                "isList": false
              }
            ]
          },
          {
            "name": "cursor",
            "isRequired": false,
            "isNullable": false,
            "inputTypes": [
              {
                "type": "PostWhereUniqueInput",
                "namespace": "prisma",
                "location": "inputObjectTypes",
                "isList": false
              }
            ]
          },
          {
            "name": "take",
            "isRequired": false,
            "isNullable": false,
            "inputTypes": [
              {
                "type": "Int",
                "location": "scalar",
                "isList": false
              }
            ]
          },
          {
            "name": "skip",
            "isRequired": false,
            "isNullable": false,
            "inputTypes": [
              {
                "type": "Int",
                "location": "scalar",
                "isList": false
              }
            ]
          },
          {
            "name": "distinct",
            "isRequired": false,
            "isNullable": false,
            "inputTypes": [
              {
                "type": "PostScalarFieldEnum",
                "namespace": "prisma",
                "location": "enumTypes",
                "isList": true
              }
            ]
          }
        ],
        "isNullable": true,
        "outputType": {
          "type": "Post",
          "namespace": "model",
          "location": "outputObjectTypes",
          "isList": true
        }
      },
      {
        "name": "keywords",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "String",
          "location": "scalar",
          "isList": true
        }
      },
      {
        "name": "biography",
        "args": [],
        "isNullable": false,
        "outputType": {
          "type": "Json",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "_count",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "UserCountOutputType",
          "namespace": "prisma",
          "location": "outputObjectTypes",
          "isList": false
        }
      }
    ]
  },
  "isInputType": false
}
