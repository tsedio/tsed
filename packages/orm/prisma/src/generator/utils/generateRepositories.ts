import {DMMF} from "@prisma/generator-helper";
import {toMap} from "@tsed/core";
import {camelCase, pascalCase} from "change-case";
import path from "path";
import pluralize from "pluralize";
import {ClassDeclaration, Project, Scope} from "ts-morph";
import {DmmfModel} from "../domain/DmmfModel.js";
import {generateOutputsBarrelFile} from "./generateOutputsBarrelFile.js";

interface MethodOptions {
  repository: ClassDeclaration;
  name: string;
  model: string;
  returnType?: string | undefined;
  hasQuestionToken?: boolean;
  isAsync?: boolean;
}

function addDelegatedMethod({name, hasQuestionToken, repository, model, returnType, isAsync = true}: MethodOptions) {
  const method = repository.addMethod({
    name: name,
    isAsync,
    returnType: returnType ? `Promise<${returnType}>` : undefined,
    parameters: [
      {
        name: "args",
        type: `Prisma.${model}${pascalCase(name)}Args`,
        hasQuestionToken
      }
    ]
  });

  if (returnType) {
    method.setBodyText(`const obj = await this.collection.${name}(args);
        return this.deserialize<${returnType}>(obj);`);
  } else {
    method.setBodyText(`return this.collection.${name}(args)`);
  }
}

export function generateRepositories(dmmf: DMMF.Document, project: Project, baseDirPath: string) {
  const modelsMap = toMap<string, DMMF.Model>(dmmf.datamodel.models, "name");
  const models = DmmfModel.getModels(dmmf, modelsMap);
  const repoDirPath = path.resolve(baseDirPath, "repositories");
  const repoDirectory = project.createDirectory(repoDirPath);
  const repositoriesIndex = repoDirectory.createSourceFile(`index.ts`, undefined, {overwrite: true});

  const exportedModels = models.map((model) => {
    const name = pascalCase(`${pluralize(model.name)}Repository`);
    const modelName = model.toString();
    const sourceFile = repoDirectory.createSourceFile(`${name}.ts`, undefined, {overwrite: true});

    sourceFile.addImportDeclarations([
      {
        moduleSpecifier: "@tsed/core",
        namedImports: ["isArray"]
      },
      {
        moduleSpecifier: "@tsed/json-mapper",
        namedImports: ["deserialize"]
      },
      {
        moduleSpecifier: "@tsed/di",
        namedImports: ["Injectable", "Inject"]
      },
      {
        moduleSpecifier: "../services/PrismaService",
        namedImports: ["PrismaService"]
      },
      {
        moduleSpecifier: "../client",
        namedImports: ["Prisma", model.name]
      },
      {
        moduleSpecifier: "../models",
        namedImports: [model.toString()]
      }
    ]);

    const repository = sourceFile.addClass({
      name,
      isExported: true,
      decorators: [
        {
          name: "Injectable",
          arguments: []
        }
      ]
    });

    repository.addProperty({
      name: "prisma",
      type: "PrismaService",
      scope: Scope.Protected,
      decorators: [
        {
          name: "Inject",
          arguments: []
        }
      ]
    });

    repository
      .addGetAccessor({
        name: "collection"
      })
      .setBodyText(`return this.prisma.${camelCase(model.name)}`);

    repository
      .addGetAccessor({
        name: "groupBy"
      })
      .setBodyText(`return this.collection.groupBy.bind(this.collection)`);

    repository
      .addMethod({
        name: "deserialize",
        scope: Scope.Protected,
        returnType: "T",
        parameters: [
          {
            name: "obj",
            type: `null | ${model.name} | ${model.name}[]`
          }
        ],
        typeParameters: ["T"]
      })
      .setBodyText(`return deserialize<T>(obj, {type: ${modelName}, collectionType: isArray(obj) ? Array : undefined})`);

    addDelegatedMethod({
      repository,
      name: "findUnique",
      model: model.name,
      returnType: `${modelName} | null`
    });

    addDelegatedMethod({
      repository,
      name: "findFirst",
      model: model.name,
      returnType: `${modelName} | null`
    });

    addDelegatedMethod({
      repository,
      name: "findMany",
      model: model.name,
      returnType: `${modelName}[]`,
      hasQuestionToken: true
    });

    addDelegatedMethod({
      repository,
      name: "create",
      model: model.name,
      returnType: `${modelName}`
    });

    addDelegatedMethod({
      repository,
      name: "update",
      model: model.name,
      returnType: `${modelName}`
    });

    addDelegatedMethod({
      repository,
      name: "upsert",
      model: model.name,
      returnType: `${modelName}`
    });

    addDelegatedMethod({
      repository,
      name: "delete",
      model: model.name,
      returnType: `${modelName}`
    });

    addDelegatedMethod({
      repository,
      name: "deleteMany",
      model: model.name,
      isAsync: false
    });

    addDelegatedMethod({
      repository,
      name: "updateMany",
      model: model.name,
      isAsync: false
    });

    addDelegatedMethod({
      repository,
      name: "aggregate",
      model: pascalCase(model.name),
      isAsync: false
    });

    return name;
  });

  generateOutputsBarrelFile(repositoriesIndex, exportedModels);
}
