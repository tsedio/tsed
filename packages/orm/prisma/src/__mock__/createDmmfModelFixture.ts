import {DmmfModel} from "../generator/domain/DmmfModel.js";
import dmmfUserModel from "./dmmfUserModel.json";

export function createDmmfModelFixture() {
  return new DmmfModel(dmmfUserModel as any);
}
