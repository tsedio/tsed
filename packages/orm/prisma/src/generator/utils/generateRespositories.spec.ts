import {createProjectFixture} from "../../__mock__/createProjectFixture";
import {createDmmfFixture} from "../../__mock__/createDmmfFixture";
import {generateRepositories} from "./generateRepositories";

describe("generateRepositories", () => {
  it("should generate repositories", () => {
    const {project, render, baseDir} = createProjectFixture("generate_repositories");
    const dmmf = createDmmfFixture();

    generateRepositories(dmmf, project, baseDir);

    const userRepo = render("/repositories/UsersRepository.ts");

    userRepo.not.toContain("Post");
    userRepo.toEqualSnapshot();

    const postsRepository = render("/repositories/PostsRepository.ts");

    postsRepository.not.toContain("User");
    postsRepository.toEqualSnapshot();
  });
});
