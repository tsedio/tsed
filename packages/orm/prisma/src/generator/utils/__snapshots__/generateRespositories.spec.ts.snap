// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateRepositories > should generate repositories (posts) 1`] = `
"import { isArray } from "@tsed/core";
import { deserialize } from "@tsed/json-mapper";
import { Injectable, Inject } from "@tsed/di";
import { PrismaService } from "../services/PrismaService";
import { Prisma, Post } from "../client/index";
import { PostModel } from "../models/index";

@Injectable()
export class PostsRepository {
  @Inject()
  protected prisma: PrismaService;

  get collection() {
    return this.prisma.post
  }

  get groupBy() {
    return this.collection.groupBy.bind(this.collection)
  }

  protected deserialize<T>(obj: null | Post | Post[]): T {
    return deserialize<T>(obj, { type: PostModel, collectionType: isArray(obj) ? Array : undefined })
  }

  async findUnique(args: Prisma.PostFindUniqueArgs): Promise<PostModel | null> {
    const obj = await this.collection.findUnique(args);
    return this.deserialize<PostModel | null>(obj);
  }

  async findFirst(args: Prisma.PostFindFirstArgs): Promise<PostModel | null> {
    const obj = await this.collection.findFirst(args);
    return this.deserialize<PostModel | null>(obj);
  }

  async findMany(args?: Prisma.PostFindManyArgs): Promise<PostModel[]> {
    const obj = await this.collection.findMany(args);
    return this.deserialize<PostModel[]>(obj);
  }

  async create(args: Prisma.PostCreateArgs): Promise<PostModel> {
    const obj = await this.collection.create(args);
    return this.deserialize<PostModel>(obj);
  }

  async update(args: Prisma.PostUpdateArgs): Promise<PostModel> {
    const obj = await this.collection.update(args);
    return this.deserialize<PostModel>(obj);
  }

  async upsert(args: Prisma.PostUpsertArgs): Promise<PostModel> {
    const obj = await this.collection.upsert(args);
    return this.deserialize<PostModel>(obj);
  }

  async delete(args: Prisma.PostDeleteArgs): Promise<PostModel> {
    const obj = await this.collection.delete(args);
    return this.deserialize<PostModel>(obj);
  }

  deleteMany(args: Prisma.PostDeleteManyArgs) {
    return this.collection.deleteMany(args)
  }

  updateMany(args: Prisma.PostUpdateManyArgs) {
    return this.collection.updateMany(args)
  }

  aggregate(args: Prisma.PostAggregateArgs) {
    return this.collection.aggregate(args)
  }
}
"
`;

exports[`generateRepositories > should generate repositories (user) 1`] = `
"import { isArray } from "@tsed/core";
import { deserialize } from "@tsed/json-mapper";
import { Injectable, Inject } from "@tsed/di";
import { PrismaService } from "../services/PrismaService";
import { Prisma, User } from "../client/index";
import { UserModel } from "../models/index";

@Injectable()
export class UsersRepository {
  @Inject()
  protected prisma: PrismaService;

  get collection() {
    return this.prisma.user
  }

  get groupBy() {
    return this.collection.groupBy.bind(this.collection)
  }

  protected deserialize<T>(obj: null | User | User[]): T {
    return deserialize<T>(obj, { type: UserModel, collectionType: isArray(obj) ? Array : undefined })
  }

  async findUnique(args: Prisma.UserFindUniqueArgs): Promise<UserModel | null> {
    const obj = await this.collection.findUnique(args);
    return this.deserialize<UserModel | null>(obj);
  }

  async findFirst(args: Prisma.UserFindFirstArgs): Promise<UserModel | null> {
    const obj = await this.collection.findFirst(args);
    return this.deserialize<UserModel | null>(obj);
  }

  async findMany(args?: Prisma.UserFindManyArgs): Promise<UserModel[]> {
    const obj = await this.collection.findMany(args);
    return this.deserialize<UserModel[]>(obj);
  }

  async create(args: Prisma.UserCreateArgs): Promise<UserModel> {
    const obj = await this.collection.create(args);
    return this.deserialize<UserModel>(obj);
  }

  async update(args: Prisma.UserUpdateArgs): Promise<UserModel> {
    const obj = await this.collection.update(args);
    return this.deserialize<UserModel>(obj);
  }

  async upsert(args: Prisma.UserUpsertArgs): Promise<UserModel> {
    const obj = await this.collection.upsert(args);
    return this.deserialize<UserModel>(obj);
  }

  async delete(args: Prisma.UserDeleteArgs): Promise<UserModel> {
    const obj = await this.collection.delete(args);
    return this.deserialize<UserModel>(obj);
  }

  deleteMany(args: Prisma.UserDeleteManyArgs) {
    return this.collection.deleteMany(args)
  }

  updateMany(args: Prisma.UserUpdateManyArgs) {
    return this.collection.updateMany(args)
  }

  aggregate(args: Prisma.UserAggregateArgs) {
    return this.collection.aggregate(args)
  }
}
"
`;

exports[`generateRepositories should generate repositories (posts) 1`] = `
"import { isArray } from \\"@tsed/core\\";
import { deserialize } from \\"@tsed/json-mapper\\";
import { Injectable, Inject } from \\"@tsed/di\\";
import { PrismaService } from \\"../services/PrismaService\\";
import { Prisma, Post } from \\"../client/index\\";
import { PostModel } from \\"../models/index\\";

@Injectable()
export class PostsRepository {
  @Inject()
  protected prisma: PrismaService;

  get collection() {
    return this.prisma.post
  }

  get groupBy() {
    return this.collection.groupBy.bind(this.collection)
  }

  protected deserialize<T>(obj: null | Post | Post[]): T {
    return deserialize<T>(obj, { type: PostModel, collectionType: isArray(obj) ? Array : undefined })
  }

  async findUnique(args: Prisma.PostFindUniqueArgs): Promise<PostModel | null> {
    const obj = await this.collection.findUnique(args);
    return this.deserialize<PostModel | null>(obj);
  }

  async findFirst(args: Prisma.PostFindFirstArgs): Promise<PostModel | null> {
    const obj = await this.collection.findFirst(args);
    return this.deserialize<PostModel | null>(obj);
  }

  async findMany(args?: Prisma.PostFindManyArgs): Promise<PostModel[]> {
    const obj = await this.collection.findMany(args);
    return this.deserialize<PostModel[]>(obj);
  }

  async create(args: Prisma.PostCreateArgs): Promise<PostModel> {
    const obj = await this.collection.create(args);
    return this.deserialize<PostModel>(obj);
  }

  async update(args: Prisma.PostUpdateArgs): Promise<PostModel> {
    const obj = await this.collection.update(args);
    return this.deserialize<PostModel>(obj);
  }

  async upsert(args: Prisma.PostUpsertArgs): Promise<PostModel> {
    const obj = await this.collection.upsert(args);
    return this.deserialize<PostModel>(obj);
  }

  async delete(args: Prisma.PostDeleteArgs): Promise<PostModel> {
    const obj = await this.collection.delete(args);
    return this.deserialize<PostModel>(obj);
  }

  deleteMany(args: Prisma.PostDeleteManyArgs) {
    return this.collection.deleteMany(args)
  }

  updateMany(args: Prisma.PostUpdateManyArgs) {
    return this.collection.updateMany(args)
  }

  aggregate(args: Prisma.PostAggregateArgs) {
    return this.collection.aggregate(args)
  }
}
"
`;

exports[`generateRepositories should generate repositories (user) 1`] = `
"import { isArray } from \\"@tsed/core\\";
import { deserialize } from \\"@tsed/json-mapper\\";
import { Injectable, Inject } from \\"@tsed/di\\";
import { PrismaService } from \\"../services/PrismaService\\";
import { Prisma, User } from \\"../client/index\\";
import { UserModel } from \\"../models/index\\";

@Injectable()
export class UsersRepository {
  @Inject()
  protected prisma: PrismaService;

  get collection() {
    return this.prisma.user
  }

  get groupBy() {
    return this.collection.groupBy.bind(this.collection)
  }

  protected deserialize<T>(obj: null | User | User[]): T {
    return deserialize<T>(obj, { type: UserModel, collectionType: isArray(obj) ? Array : undefined })
  }

  async findUnique(args: Prisma.UserFindUniqueArgs): Promise<UserModel | null> {
    const obj = await this.collection.findUnique(args);
    return this.deserialize<UserModel | null>(obj);
  }

  async findFirst(args: Prisma.UserFindFirstArgs): Promise<UserModel | null> {
    const obj = await this.collection.findFirst(args);
    return this.deserialize<UserModel | null>(obj);
  }

  async findMany(args?: Prisma.UserFindManyArgs): Promise<UserModel[]> {
    const obj = await this.collection.findMany(args);
    return this.deserialize<UserModel[]>(obj);
  }

  async create(args: Prisma.UserCreateArgs): Promise<UserModel> {
    const obj = await this.collection.create(args);
    return this.deserialize<UserModel>(obj);
  }

  async update(args: Prisma.UserUpdateArgs): Promise<UserModel> {
    const obj = await this.collection.update(args);
    return this.deserialize<UserModel>(obj);
  }

  async upsert(args: Prisma.UserUpsertArgs): Promise<UserModel> {
    const obj = await this.collection.upsert(args);
    return this.deserialize<UserModel>(obj);
  }

  async delete(args: Prisma.UserDeleteArgs): Promise<UserModel> {
    const obj = await this.collection.delete(args);
    return this.deserialize<UserModel>(obj);
  }

  deleteMany(args: Prisma.UserDeleteManyArgs) {
    return this.collection.deleteMany(args)
  }

  updateMany(args: Prisma.UserUpdateManyArgs) {
    return this.collection.updateMany(args)
  }

  aggregate(args: Prisma.UserAggregateArgs) {
    return this.collection.aggregate(args)
  }
}
"
`;
